toolchain/semantics/semantics_ir_formatter.cpp

```diff
@@ -246,21 +245,78 @@ class NodeNamer {
   }
 
   auto AddBlockLabel(ScopeIndex scope_idx, SemanticsNodeBlockId block_id,
-                     std::string name = "") -> void {
-    if (!block_id.is_valid()) {
+                     std::string name = "",
+                     ParseTree::Node parse_node = ParseTree::Node::Invalid)
+      -> void {
+    if (!block_id.is_valid() || labels[block_id.index].second) {
       return;
     }
 
-    ParseTree::Node parse_node = ParseTree::Node::Invalid;
-    if (auto& block = semantics_ir_.GetNodeBlock(block_id); !block.empty()) {
-      parse_node = semantics_ir_.GetNode(block.front()).parse_node();
+    if (parse_node == ParseTree::Node::Invalid) {
+      if (auto& block = semantics_ir_.GetNodeBlock(block_id); !block.empty()) {
+        parse_node = semantics_ir_.GetNode(block.front()).parse_node();
+      }
     }
 
     labels[block_id.index] = {scope_idx,
                               GetScopeInfo(scope_idx).labels.AllocateName(
                                   *this, parse_node, std::move(name))};
   }
 
+  // Finds and adds a suitable block label for the given semantics node.
+  auto AddBlockLabel(ScopeIndex scope_idx, SemanticsNodeBlockId block_id,
+                     SemanticsNode node) -> void {
+    llvm::StringRef name;
+    switch (parse_tree_.node_kind(node.parse_node())) {
+      case ParseNodeKind::IfExpressionIf:
+        switch (node.kind()) {
+          case SemanticsNodeKind::BranchIf:
+            name = "if.expr.then";
+            break;
+          case SemanticsNodeKind::Branch:
+            name = "if.expr.else";
+            break;
+          case SemanticsNodeKind::BranchWithArg:
+            name = "if.expr.cont";
+            break;
+          default:
+            break;
+        }
+        break;
+
+      case ParseNodeKind::IfCondition:
+        switch (node.kind()) {
+          case SemanticsNodeKind::BranchIf:
+            name = "if.then";
+            break;
+          case SemanticsNodeKind::Branch:
+            name = "if.else";
+            break;
+          default:
+            break;
+        }
+        break;
+
+      case ParseNodeKind::IfStatement:
+        name = "if.cont";
```

Comment:
Is the parse node the *predecessor* of the block? That wasn't obvious to me -- but not sure where to clarify it?

Reply:
The parse tree node is the node associated with the branch instruction, so in some sense yes. I've updated the function comment to say what the given semantics node means.